pkg/operator/controller/dns/controller.go

```diff
@@ -661,26 +660,30 @@ func getIbmDNSProvider(dnsConfig *configv1.DNS, creds *corev1.Secret, instanceID
 	}
 
 	providerCfg := ibm.Config{
-		APIKey:     string(creds.Data["ibmcloud_api_key"]),
-		InstanceID: instanceID,
-		Zones:      zones,
-		UserAgent:  userAgent,
+		APIKey:    string(creds.Data["ibmcloud_api_key"]),
+		Zones:     zones,
+		UserAgent: userAgent,
 	}
 
 	if isPublic {
+		providerCfg.InstanceID = instanceCRN
 		provider, err := ibmpublicdns.NewProvider(providerCfg)
 		if err != nil {
 			return nil, fmt.Errorf("failed to initialize IBM CIS DNS provider: %w", err)
 		}
 		log.Info("successfully initialized IBM CIS DNS provider")
 		return provider, nil
 	} else {
+		matches := ibm.IBMResourceCRNRegexp.FindStringSubmatch(instanceCRN)
+		if matches == nil {
+			return nil, fmt.Errorf("CRN: %s does not match expected format: %s", instanceCRN, ibm.IBMResourceCRNRegexp)
+		}
+		providerCfg.InstanceID = matches[ibm.IBMResourceCRNRegexp.SubexpIndex("guid")]
```

Comment:
I think you need to check the bound on `matches`. Looking at `ibm.IBMResourceCRNRegexp.SubexpIndex("guid")`, that can return `-1`. I think we should be more defensive here.

Reply:
As [IBMResourceCRNTegexp](https://github.com/Karthik-K-N/cluster-ingress-operator/blob/10966557ef8592858fe068065e526a87c37315c5/pkg/dns/ibm/common.go#L16-L15) is declared as constant it always expected to return 6. So do you think we need to check additionally for matches length >= 6?

Reply:
Will `matches` always have 6 entries if the lookup succeeds?

Reply:
With a valid instanceCRN the matches length will be 9,
```
        IBMResourceCRNRegexp := regexp.MustCompile(`^crn:v[0-9]:(?P<cloudName>[^:]*):(?P<cloudType>[^:]*):(?P<serviceName>[^:]*):(?P<location>[^:]*):(?P<scope>[^:]*):(?P<guid>[^:]*):(?P<resourceType>[^:]*):(?P<resourceID>[^:]*)$`)
	crn := "crn:v1:bluemix:public:internet-svcs:global:a/65b64c1f1c11160e8c2e4bbfbd893c2c:453c4cff-2ee0-4309-95f1-2e9384111b96::"
	matches := IBMResourceCRNRegexp.FindStringSubmatch(crn)
	if matches == nil {
		panic(fmt.Errorf("CRN: %s does not match expected format: %s", crn, IBMResourceCRNRegexp))
	}
	fmt.Println(IBMResourceCRNRegexp.SubexpIndex("guid")) // 6
	fmt.Println(len(matches)) // 9
        fmt.Println(matches[IBMResourceCRNRegexp.SubexpIndex("guid")]) // 453c4cff-2ee0-4309-95f1-2e9384111b96
```